lib/uefi/boot_service_provider.cpp

```diff
@@ -0,0 +1,380 @@
+#include "lib/dlmalloc.h"
+#include "types.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+EfiStatus unload(EfiHandle handle) { return SUCCESS; }
+
+bool guid_eq(const EfiGuid *a, const EfiGuid *b) {
+  return memcmp(a, b, sizeof(*a)) == 0;
+}
+
+bool guid_eq(const EfiGuid *a, const EfiGuid &b) {
+  return memcmp(a, &b, sizeof(*a)) == 0;
+}
+
+vmm_aspace_t *get_address_space() {
+  static vmm_aspace_t *aspace = nullptr;
+  if (aspace == nullptr) {
+    auto err = vmm_create_aspace(&aspace, "linux_kernel", 0);
+    if (err) {
+      printf("Failed to create address space for linux kernel %d\n", err);
+      return nullptr;
+    }
+    vmm_set_active_aspace(aspace);
+  }
+  return aspace;
+}
+
+void *identity_map(void *addr, size_t size) {
+  size = ROUNDUP(size, PAGE_SIZE);
+  auto vaddr = reinterpret_cast<vaddr_t>(addr);
+  paddr_t pa{};
+  uint flags{};
+  auto aspace = get_address_space();
+  auto err = arch_mmu_query(&aspace->arch_aspace, vaddr, &pa, &flags);
+  if (err) {
+    printf("Failed to query physical address for memory 0x%p\n", addr);
+    return nullptr;
+  }
+
+  err = arch_mmu_unmap(&aspace->arch_aspace, vaddr, size / PAGE_SIZE);
+  if (err) {
+    printf("Failed to unmap virtual address 0x%lx\n", vaddr);
+    return nullptr;
+  }
+  arch_mmu_map(&aspace->arch_aspace, pa, pa, size / PAGE_SIZE, flags);
+  if (err) {
+    printf("Failed to identity map physical address 0x%lx\n", pa);
+    return nullptr;
+  }
+  printf("Identity mapped physical address 0x%lx flags 0x%x\n", pa, flags);
+
+  return reinterpret_cast<void *>(pa);
+}
+
+void *alloc_page(size_t size, size_t align_log2) {
+  auto aspace = get_address_space();
+  void *vptr{};
+  status_t err = vmm_alloc_contiguous(aspace, "uefi_program", size, &vptr,
+                                      align_log2, 0, 0);
+  if (err) {
+    printf("Failed to allocate memory for uefi program %d\n", err);
+    return nullptr;
+  }
+  return identity_map(vptr, size);
+}
+
+void *alloc_page(void *addr, size_t size, size_t align_log2) {
+  if (addr == nullptr) {
+    return alloc_page(size, align_log2);
+  }
+  auto err =
+      vmm_alloc_contiguous(get_address_space(), "uefi_program", size, &addr,
+                           align_log2, VMM_FLAG_VALLOC_SPECIFIC, 0);
+  if (err) {
+    printf(
+        "Failed to allocate memory for uefi program @ fixed address 0x%p %d , "
+        "falling back to non-fixed allocation\n",
+        addr, err);
+    return alloc_page(size, align_log2);
+  }
+  return identity_map(addr, size);
+}
+
+EfiStatus handle_protocol(EfiHandle handle, const EfiGuid *protocol,
+                          void **intf) {
+  if (guid_eq(protocol, LOADED_IMAGE_PROTOCOL_GUID)) {
+    printf("handle_protocol(%p, LOADED_IMAGE_PROTOCOL_GUID, %p);\n", handle,
+           intf);
+    auto loaded_image = static_cast<EFI_LOADED_IMAGE_PROTOCOL *>(
+        malloc(sizeof(EFI_LOADED_IMAGE_PROTOCOL)));
+    memset(loaded_image, 0, sizeof(*loaded_image));
```

Comment:
suggestion: since in C++ can do *loaded_image = {};

Reply:
Done